### PR TITLE
add Forgejo

### PR DESCRIPTION
**Pull request by SeaswimmerTheFsh**
# Description

This pull request adds an egg for [Forgejo](https://forgejo.org/), an open-source Git hosting service forked from [Gitea](https://gitea.com/).

## Checklist for all submissions

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/parkervcp/eggs/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
* [x] Have you tested and reviewed your changes with confidence that everything works?
* [x] Did you branch your changes and PR from that branch and not from your master branch?
* [x] You verify that the start command applied does not use a shell script `I think?`
  * [ ] If some script is needed then it is part of a current yolk or a PR to add one
* [x] The egg was exported from the panel

## New egg Submissions

1. [x] Does your submission pass tests (server is connectable)?
2. [ ] Does your egg use a custom docker image?
    * [ ] Have you tried to use a generic image?
    * [ ] Did you PR the necessary changes to make it work?
3. [x] Have you added the egg to the main README.md and any other README files in subdirectories of the egg (e.g /game_eggs) according to the alphabetical order?
4. [x] Have you added a unique README.md for the egg you are adding according to the alphabetical order?
5. [x] You verify that the start command applied does not use a shell script `again, I think?`
    * [ ] If some script is needed then it is part of a current yolk or a PR to add one
6. [x] The egg was exported from the panel

### CODE REVIEW

**Comment by SeaswimmerTheFsh** [comment]
The only thing I noticed while making this is that, hypothetically, if a user deleted the binary that the installer downloads and uploaded their own (or created one in the ui), you could execute bash scripts. The gitea egg already in this repository is also vulnerable to this exploit, and I'm unsure of how to go about fixing it.
```
[Pterodactyl Daemon]: Checking server disk space usage, this could take a few seconds...
[Pterodactyl Daemon]: Updating process configuration files...
[Pterodactyl Daemon]: Ensuring file permissions are set correctly, this could take a few seconds...
container@pterodactyl~ Server marked as starting...
[Pterodactyl Daemon]: Pulling Docker container image, this could take a few minutes to complete...
Pulling from parkervcp/yolks 
Status: Image is up to date for ghcr.io/parkervcp/yolks:debian 
Digest: sha256:43d1833f9ee999fb569bc2f26d11ac19f0a8b66d5106f8ba9b6419bb8363acb7 
[Pterodactyl Daemon]: Finished pulling Docker container image
:/home/container$ ./gitea web -p ${SERVER_PORT} -c ./app.ini
successfully ran an arbitrary bash script through this egg's startup command without administrator access
container@pterodactyl~ Server marked as offline...
[Pterodactyl Daemon]: ---------- Detected server process in a crashed state! ----------
[Pterodactyl Daemon]: Exit code: 0
[Pterodactyl Daemon]: Out of memory: false
```

**Comment by QuintenQVD0** [comment]
> The only thing I noticed while making this is that, hypothetically, if a user deleted the binary that the installer downloads and uploaded their own (or created one in the ui), you could execute bash scripts. The gitea egg already in this repository is also vulnerable to this exploit, and I'm unsure of how to go about fixing it.
> 
> ```
> [Pterodactyl Daemon]: Checking server disk space usage, this could take a few seconds...
> [Pterodactyl Daemon]: Updating process configuration files...
> [Pterodactyl Daemon]: Ensuring file permissions are set correctly, this could take a few seconds...
> container@pterodactyl~ Server marked as starting...
> [Pterodactyl Daemon]: Pulling Docker container image, this could take a few minutes to complete...
> Pulling from parkervcp/yolks 
> Status: Image is up to date for ghcr.io/parkervcp/yolks:debian 
> Digest: sha256:43d1833f9ee999fb569bc2f26d11ac19f0a8b66d5106f8ba9b6419bb8363acb7 
> [Pterodactyl Daemon]: Finished pulling Docker container image
> :/home/container$ ./gitea web -p ${SERVER_PORT} -c ./app.ini
> successfully ran an arbitrary bash script through this egg's startup command without administrator access
> container@pterodactyl~ Server marked as offline...
> [Pterodactyl Daemon]: ---------- Detected server process in a crashed state! ----------
> [Pterodactyl Daemon]: Exit code: 0
> [Pterodactyl Daemon]: Out of memory: false
> ```

Their is no way. That is litterly possible with al pterodactyl eggs.

File: software/forgejo/egg-forgejo.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
...server\r\n\r\n## install forgejo\r\nif [ -z \"${VERSION}\" ] || [ \"${VERSION}\" == \"latest\" ]; then\r\n    echo -e \"Downloading Forgejo $LATEST (latest)\"\r\n    URL=https:\/\/codeberg.org\/forgejo\/forgejo\/releases\/download\/v${LATEST}\/forgejo-${LATEST}-linux-${ARCH}\r\nelse\r\n    echo -e \"Downloading Forgejo $VERSION\"\r\n    URL=https:\/\/codeberg.org\/forgejo\/forgejo\/releases\/download\/v${VERSION}\/forgejo-${VERSION}-linux-${ARCH}\r\nfi\r\n\r\necho -e \"Download URL: $URL\"\r\n\r\ncurl -sSL -o forgejo $URL\r\n\r\nchmod +x forgejo\r\n\r\nmkdir -p custom\r\n\r\nif [ -f \"\/mnt\/server\/custom\/app.ini\" ]; then\r\n    echo \"Config file exists\"\r\nelse\r\n    echo \"[server]\r\n    LOCAL_ROOT_URL = http:\/\/${SERVER_IP}:${SERVER_PORT}\/\r\n    DOMAIN           = ${SERVER_IP}\r\n    HTTP_PORT        = ${SERVER_PORT}\r\n    ROOT_URL         = http:\/\/${SERVER_IP}:${SERVER_PORT}\/\r\n    DISABLE_SSH      = ${DISABLE_SSH}\r\n    SSH_PORT         = ${SSH_PORT}\" > \/mnt\/server\/custom\/app.ini\r\nfi\r\n\r\necho -e \"-------------------------------------------------\"\r\necho -e \"Installation completed\"\r\necho -e \"-------------------------------------------------\"",
```
Comment (QuintenQVD0) [inline software/forgejo/egg-forgejo.json:25]: if you are going to mention a file, please mention it on the master branch

File: software/forgejo/egg-forgejo.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            "container": "debian:bullseye-slim",
```
Comment (QuintenQVD0) [inline software/forgejo/egg-forgejo.json:26]: this can be parker debian installer image as we in the future will move to debian 12

**Review by QuintenQVD0 — CHANGES_REQUESTED** [review]

File: software/forgejo/egg-forgejo.json (hunk 2 above)
Comment (SeaswimmerTheFsh) [reply software/forgejo/egg-forgejo.json]: if by this you mean `ghcr.io/parkervcp/yolks:debian`, this is changed in the latest commit

File: software/forgejo/egg-forgejo.json (hunk 1 above)
Comment (SeaswimmerTheFsh) [reply software/forgejo/egg-forgejo.json]: changed in latest commit

**Comment by SeaswimmerTheFsh** [comment]
> Their is no way. That is litterly possible with al pterodactyl eggs.
Oh I didn't realize that, sorry.

Thanks for the review, I've made the requested changes in the most recent commit.

**Comment by QuintenQVD0** [comment]
I appended a commit that adds the main binary to the file_denylist, it should prevent user overwriting but is still experimental.

**Review by QuintenQVD0 — APPROVED** [review]
LGTM!

**Pull request merged**